docs/content/sips/xxx-trigger-executors.md

```diff
@@ -0,0 +1,65 @@
+title = "SIP xxx - Trigger Executors"
+template = "main"
+
+---
+
+Summary: A new architecture for Spin trigger executors.
+
+Owner: lann.martin@fermyon.com
+
+Created: April 29, 2022
+
+<!-- Updated: April 4, 2022 -->
+
+## Background
+
+Spin currently includes built-in implementations for two trigger types: HTTP, and Redis (PubSub).
+We want to allow for additional trigger types in the future, whether they are "internal" types developed and
+supported within the Spin codebase, or "external" types developed by the broader community.
+
+While it is currently possible to build external trigger types - as in the
+[TimerTrigger example](https://github.com/fermyon/spin/blob/main/examples/spin-timer/src/main.rs) - it requires replicating much of the
+functionality provided by the `spin up` command.
+
+## Proposal
+
+### Trigger types are implemented as "trigger executors"
+
+These executors are regular subcommands/binaries, but are not normally executed directly by
+users.
+
+* "Internal" triggers use `spin` subcommands, e.g. `spin trigger http`
+* "External" triggers use external binaries, e.g. `spin-trigger-timer`
+* Trigger executors share common code, including common CLI options in a `spin-trigger` crate
+
+### The `spin up` command changes
+
+It remains the primary entrypoint for starting a Spin app, but its behavior is changed to:
+
+* Detect the trigger type for an application
+* Find a trigger executor for that type (e.g. check the `PATH` for `"spin-trigger-${type}"`)
+* `exec` the trigger executor, forwarding its own CLI arguments
```

Comment:
What new concerns exist with two processes collaborating?  If the parent (spin up) process crashes, what happens to the trigger?  E.g. can the user determine which trigger process corresponds to a crashed spin up in order to kill it?  What is the flow when the user runs `spin up` and then hits Ctrl+C - does that kill the trigger process and return to the spin process, or does it kill the spin process and we need to forward it to the trigger process?  If the former, can the spin process tell the difference between "user killed" and "program crashed" - or do we not need to care?

Reply:
In Linux, a parent process dying will normally kill all children automatically. Worth review on other platforms, but I would expect to be able to implement that behavior if it isn't the default.

Reply:
There's no equivalent on Windows either - you always end up spawning a child process.